### PR TITLE
Enable Block Renaming support for (almost) all blocks

### PR DESCRIPTION
**Pull request by getdave**
## What?

Following on from https://github.com/WordPress/gutenberg/pull/53735, this PR enables the ability to rename **all blocks** by _default_, with the exception of the following blocks:

- `core/block`
- `core/template-part`
- `core/pattern`
- `core/navigation`

## Why?

Because it was suggested we should open up this feature.

## How?

- Adds metadata to all blocks by default (the `__experimentalMetadata` support is removed).
- Add new `__experimentalBlockRenaming` block supports.
- Set excluded blocks support to `false`.
- Provides custom label definitions for those blocks that already implement one.

## Testing Instructions

- Add any block (except for the exceptions defined above).
- Open list view
- Click ellipsis menu and click `Rename`.
- Enter a name.
- See block uses custom name.
- Repeat with all of the exception blocks and verify they cannot be renamed.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/444434/8e8dac7a-5f1f-4678-93ad-d5d81821849d

### CODE REVIEW

**Comment by getdave** [comment]
I am unsure as to how this might impact accessibility. Specifically I notice that the paragraph block uses a custom label for the block when the context is `accessibility`. I do not know what that is used for and how to test for it. Some guidance here would be appreciated from those experienced with a11y (cc @jeryj).

**Comment by getdave** [comment]
I think we may need a new API to retrieve the block's custom name. Sprinkling `attributes.metadata.name` everywhere seems like it could get messy fast.

**Comment by jeryj** [comment]
> Specifically I notice that the paragraph block uses a custom label for the block when the context is accessibility. I do not know what that is used for and how to test for it. 

Could you explain this part more? I'm unfamiliar with it. Could you point me in the right direction of how to look into this? Thanks, Dave!

**Comment by getdave** [comment]
> > Specifically I notice that the paragraph block uses a custom label for the block when the context is accessibility. I do not know what that is used for and how to test for it.
> 
> Could you explain this part more? I'm unfamiliar with it. Could you point me in the right direction of how to look into this? Thanks, Dave!

👍 It's here

https://github.com/WordPress/gutenberg/blob/eabcc9902c0a0eb2c58f2f37ac4140c00c1081d9/packages/block-library/src/paragraph/index.js#L30-L41

**Comment by draganescu** [comment]
I love this. @WordPress/gutenberg-design which blocks should support manual labeling in the list view? All of them?

@getdave I think not having another abtractisation is actually good because this is an override that feels natural to read in code as this is coming from this other place, instead of magicSpell() -> label.

**Comment by scruffian** [comment]
If we're going to make this available to all blocks maybe we don't need a block support at all.

**Comment by richtabor** [comment]
I mentioned this on https://github.com/WordPress/gutenberg/pull/50135#issuecomment-1712366161, but I did expect the placeholder text to be the heading block's auto-label (whatever the content is for the heading block) instead of maintaining the block name. 

Flow: 
- Add a heading block with text of "Learn more"
- Rename the heading block to "CTA Heading"
- Press save, closing the modal.
- Open rename modal for the heading.
- Remove the "CTA Heading" value, to reveal the placeholder. 
- See "Heading" placeholder, instead of "Learn more" (which is what the heading label will be when I press save).

**Comment by richtabor** [comment]
How do we get "Block Name" added after the existing fields in Advanced? 

CleanShot 2023-09-15 at 15 59 16

**Comment by richtabor** [comment]
> If we're going to make this available to all blocks maybe we don't need a block support at all.

Is there a case for why we wouldn't make this globally available?

**Comment by getdave** [comment]
> > If we're going to make this available to all blocks maybe we don't need a block support at all.
> 
> Is there a case for why we wouldn't make this globally available?

Wrapper blocks - template parts, patterns...etc. It might be unusual if you could rename these? How would we like to handle those?

**Comment by richtabor** [comment]
> Wrapper blocks - template parts, patterns...etc. It might be unusual if you could rename these?

Well, you can rename custom template parts (not those provided by a theme). But perhaps we could just omit those wrapper blocks. 

CleanShot 2023-09-15 at 17 20 04

**Comment by jameskoster** [comment]
@richtabor I suppose the question is whether renaming a synced block in a local context should create an alias or update the source. Updating the source seems fine to me, probably with a Snackbar confirmation to clarify the change was global.

**Comment by draganescu** [comment]
I think updating reusable / synced elements _global names_ should not work from the list view. The list view is _local_ - it would be OK in my head if reusable / synced elements would have their global label and a local label. If you want to change the global label a modal action is better IMO so it signifies the impact of the action.

The idea is to not rename Testimonials synced pattern to talking-heads when building a layout and then panic in the patterns library that you lost Testimonials b/c you didn't see the snackbar.

**Comment by jameskoster** [comment]
> If you want to change the global label a modal action is better IMO so it signifies the impact of the action.

That's part of the issue, renaming from list view invokes a modal. So in the case of synced blocks it might not be clear whether you're updating globally or adding a local alias, especially as Rename actions elsewhere are global. 

I suppose the first question to answer is whether or not to allow aliases for synced blocks. Partially synced should be considered too. It might be better for that discussion to happen elsewhere to avoid holding up this PR?

**Comment by richtabor** [comment]
> @richtabor I suppose the question is whether renaming a synced block in a local context should create an alias or update the source. 

> The idea is to not rename Testimonials synced pattern to talking-heads when building a layout and then panic in the patterns library that you lost Testimonials b/c you didn't see the snackbar.

I lean towards local-only as well. 

The other rename flows are modals as well, but it's clear that you're renaming a pattern/part — rather than a block name/list view. But I'm ok omitting template parts and synced patterns from rename. 

This traces back to the notion of whether you're affecting every instance of it or not (like invoking a focus/isolated view).

**Comment by getdave** [comment]
So the consensus is make it conditionally available to all blocks save for template parts, patterns...etc?

**Comment by jameskoster** [comment]
> But I'm ok omitting template parts and synced patterns from rename

I'd lean towards omitting. 

1. It remains unclear how useful aliases are for synced blocks, especially as they _generally_ have intuitive names to begin with.
2. There's potential for confusion as to whether the change would be perceived as local or global.

Seems fine to look into this if a need arises, rather than introducing potential headaches ahead of time.

**Comment by getdave** [comment]
Ok now on to the technical question of how best to do this.

Currently support for the feature rests on the block adding support for `__experimentalMetadata` and then specifically for the `name` portion of that feature.

We could:

- enable metadata support for _all_ blocks.
- make the rename support conditional on a new block support (`__experimentalBlockRenaming`).
- make the block support truthy (can we do this!?) by default.
- add `__experimentalBlockRenaming: false` to the `block.json` of all the block swe've decided (see above) we _won't_ support.

Any prior art for such a thing?

**Comment by draganescu** [comment]
I think @scruffian 's comment above makes sense, in that if all blocks support rename, it's not an opt in and therefore it's not required to create a new block support. So we only need to make sure all blocks work - by default - with metadata and they all have the name metadata available.

**Comment by getdave** [comment]
- Move metadata [_registration_ out of experimental](https://github.com/WordPress/gutenberg/blob/191fd0f84223a543af0a2d0893567f558a9e547d/lib/experimental/blocks.php#L88-L102).
- Enable for all blocks [by removing the Core filter](https://github.com/WordPress/gutenberg/blob/191fd0f84223a543af0a2d0893567f558a9e547d/packages/block-editor/src/hooks/metadata.js#L11-L13).

File: packages/block-editor/src/hooks/block-rename-ui.js
Comment (scruffian) [inline packages/block-editor/src/hooks/block-rename-ui.js]:
```suggestion
			true //default value
```

File: packages/block-editor/src/hooks/metadata.js
Comment (getdave) [inline packages/block-editor/src/hooks/metadata.js:44]: This change ~enables metadata support~ adds the metadata attribute ~for~ to _all_ blocks.

**Comment by getdave** [comment]
Pinging @ntsekouras as this is making the block `metadata` a supported API and a default attribute on _all_ blocks.

Originally we avoided releasing this feature in https://github.com/WordPress/gutenberg/pull/47791.

File: lib/blocks.php
Comment (getdave) [inline lib/blocks.php:476]: Specifically @ntsekouras this is where it is being promoted to non-experimental. This doesn't actually change much other than it's no longer in the `experimental` directory.

**Comment by priethor** [comment]
>> Is there a case for why we wouldn't make this globally available?
>
> Wrapper blocks - template parts, patterns...etc. It might be unusual if you could rename these? How would we like to handle those?

I guess these blocks are specific enough that all blocks are allowed to be renamed except for these. Reframing Rich's question, is there any case for a **non-core** block that shouldn't allow renaming? 🤔

**Comment by getdave** [comment]
> How do we get "Block Name" added after the existing fields in Advanced?

We don't currently have a means to achieve this. This is a Slot which we fill. The ordering of fills for a given Slot is something that's been discussed much but I don't believe it ever got completed 😬 

Some more discussion [here](https://github.com/WordPress/gutenberg/issues/44024#issuecomment-1478944393).

**Comment by getdave** [comment]
> is there any case for a **non-core** block that shouldn't allow renaming?

Maybe if I invented some block that had it's own "Rename" type feature which renamed some kind of associated entity? In that case I might want to opt out of the feature.

When we add something we should probably provide a way to opt out if necessary.

The supports is experimental anyway so we could always remove it in the future if we decided it wasn't necessary.

**Comment by draganescu** [comment]
We may introduce in the future the posibility to create a reusable block - like template or pattern - via plugins. That is when opting out will be a need - but since it's opting OUT we can introduce this later when / if the need appears?

**Comment by priethor** [comment]
> but since it's opting OUT we can introduce this later when / if the need appears?

Sorry if I'm missing something here, but isn't that already achieved by setting the default value of `__experimentalBlockRenaming` to true?

**Comment by draganescu** [comment]
hi @priethor - I've given this more throught and I think @getdave is correct: we need to have opt out from the start. All blocks support renaming, except for a few blocks in core. But once the feature is out some blocks may want to not be renamed, to to have their own custom rename process (maybe for branding reasons - always appent Brand before block name?).

**Comment by getdave** [comment]
So the approach in this PR is correct. We just need a technical review on:

- expanding metadata support to 1st party API (note: we cannot do this in a separate PR as it would break block renaming for Group block).
- the new approach to block supports for block renaming

File: lib/blocks.php
Comment (ntsekouras) [reply lib/blocks.php:476]: I think having the `metadata` support is okay and has been discussed in many other issues, so it could be helpful. Should we add it in a different php file under `block-supports` folder though? 

Additionally I believe we should update the [support](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/) docs.

**Comment by ntsekouras** [comment]
>The supports is experimental anyway so we could always remove it in the future if we decided it wasn't necessary.

I'm checking the code now, but for start there is no way currently to have experimental block.json properties AFAIK. So this means the removal would not be possible if lands in core. --cc @adamziel . If the above stands we'll not need the experimental prefix.

My first thoughts though were why would we need this for every block. IMO it makes sense for some grouping blocks, but personally I don't find it useful for lots of blocks and it adds more noise. I thought the direction was for some blocks to help the user differentiate them with other means - like we do for Heading, Image blocks, etc..

**Comment by getdave** [comment]
> I'm checking the code now, but for start there is no way currently to have experimental block.json properties AFAIK. So this means the removal would not be possible if lands in core. --cc @adamziel . If the above stands we'll not need the experimental prefix.

Thanks for the review. Removing the experimental prefix is simple enough. I can do that.

> My first thoughts though were why would we need this for every block.

I'll defer that to others, but various folks have suggested to me that it would be useful to have this feature for other block types. Our TT4 release lead @MaggieCabrera specifically indicated it would be helpful for Themes.

One use case I have is Query blocks, where you would want to rename the block to remind yiou about the query type without having to inspect the query block settings. There are other blocks where it _might_ be less useful, but as this feature is not prominent it will do little harm and still allow use to cater for edge cases.

File: lib/blocks.php
Comment (getdave) [reply lib/blocks.php:476]: Yeh I was wondering that. However this is no longer a block _supports_ is it? It's just a feature of every block. 

If so then maybe it should be somewhere else in a more fundamental place lower down in the code?

**Review by ntsekouras — COMMENTED** [review]
Code wise I think this is good in general and really close. Thank you @getdave!

**Comment by adamziel** [comment]
> there is no way currently to have experimental block.json properties AFAIK

@ntsekouras you're right.

**Comment by MaggieCabrera** [comment]
I could see this being useful in any number of ways: first of all patterns. Naming patterns in the list view will separate sections of the website effectively and patterns are to always wrapped in a group block. Sometimes they are wrapped in a columns block or in the case of query block patters, the query block may not be wrapped at all.

Another good use case are layouts. Say you have a layout with a main tag (your a11y landmark), within a columns block and then a sidebar (another landmark). Naming these blocks becomes incredibly helpful to the user to help them understand how their theme is built and how best to extend it.

File: docs/reference-guides/core-blocks.md
Comment (getdave) [inline docs/reference-guides/core-blocks.md]: I do not understand the changes to these files. They were created when I ran `npm run docs:build` but they seem totally wrong. 

Why would some blocks be marked as included `blockRenaming` support and not others.

Moreover, some blocks such as Pattern which do not support blockRenaming (i.e. it's set to `false` in its `block.json`) show up as having support.

File: lib/blocks.php
Comment (getdave) [reply lib/blocks.php:476]: @ntsekouras I wonder whether infact it would be better to register the block `metadata` _attribure_ as a global attribute available to every block such as seems to have been done for `lock`.

https://github.com/WordPress/wordpress-develop/blob/a80165b78f2367f92e2eff679c4e035d1c5bdc7c/src/wp-includes/class-wp-block-type.php#L248-L250

Maybe @Mamaduka can shed some light on the best approach here? 🙏 

Remember we are discussing the `metadata` attribute and not the block support for `blockRenaming`. I say this because it's a confusing PR in that regard 😅

Comment (ntsekouras) [reply lib/blocks.php:476]: Global attributes seems a good place for this, yes.

File: packages/block-library/src/paragraph/index.js
Comment (ntsekouras) [inline packages/block-library/src/paragraph/index.js:33]: Is this something we only need for `list view`? I'm wondering if it should be part of helper hooks like `useBlockDisplayInformation` or `useBlockDisplayTitle`, or even `getBlockLabel` function etc..

There might be nuances I'm missing about why/where we use `__experimentalLabel`.

Comment (getdave) [reply packages/block-library/src/paragraph/index.js:33]: For now the output of the custom name is only utilised in the List View.

If in future there is a use case for utilising the custom name _outside_ of List View then what you recommend would be ideal 👍

File: docs/reference-guides/core-blocks.md
Comment (scruffian) [reply docs/reference-guides/core-blocks.md]: I think the `~~blockRenaming~~` syntax means that the block doesn't support that thing (notice that it's the same for `~~html~~` which is also set to false in the block.

**Review by scruffian — APPROVED** [review]
I think this is looking great, lets bring it in.

**Comment by getdave** [comment]
@scruffian Before this is merged do you think we ought to try and address @ntsekouras's [feedback regarding where to add the global metadata attribute](https://github.com/WordPress/gutenberg/pull/54426#discussion_r1341059891)? Or we could address that in an immediate follow up.

File: lib/blocks.php
Comment (Mamaduka) [reply lib/blocks.php:476]: Sorry, missed the ping.

If the attribute should be available for every block, then using a global attribute makes sense.

Here's the core PR: https://github.com/WordPress/wordpress-develop/pull/2615, if you're looking for more context.

P.S. I think we can drop the `block` prefix from the support flag and use `renaming`. I would double-check the other flags for consistency.

**Comment by getdave** [comment]
I looked into this.

As per https://github.com/WordPress/wordpress-develop/pull/2615 I suggest we merge this PR **as is** (that's inline with approach in https://github.com/WordPress/gutenberg/pull/40468) and _then_ raise a Core PR to backport the change.

That can happen in 6.5 release cycle. 

I don't believe we _need_ to backport this PR unless @priethor or other release leads want it to land in 6.4 alongside the renaming for Group blocks.

**Comment by getdave** [comment]
Rebase with aim of getting tests to ✅ as the failures seem unrelated.

**Comment by draganescu** [comment]
I agree with @getdave that this can be refined in the 6.5 cycle in the plugin. It's not a feature destined to long tail end users, however useful, and power users such as professional designers do have the plugin to use. Therefore it can stay out of core for another release imo. since we're still finding out the best ways to implement,

**Comment by getdave** [comment]
Ok thanks. I agree. 

Also just to note that to ship in Core requires

- enabling a new attribute for all blocks
- enabling a new UI feature for all blocks

This includes 3rd party blocks. So we should probably exercise caution and allow for another release cycle here.

**Comment by getdave** [comment]
I think unit test failures _might_ be legit 😕

**Comment by Mamaduka** [comment]
Finalizing the block support flag before this ships in the plugin would be nice. I think it doesn't have an `__experimental` prefix anymore.

I personally think using `block` in the name isn't needed. The context is apparent and it seems unnecessarily verbose.

Edit: Sorry, @getdave, I missed your latest commit while writing this comment 😄

**Comment by annezazu** [comment]
[Just wanted to connect some feedback](https://make.wordpress.org/test/2023/09/18/fse-program-testing-call-26-final-touches/#comment-3189) from the FSE Outreach Program's Final Touches call for testing that included the rename option with the Group block:

>Moving forward, I understand from this issue https://github.com/WordPress/gutenberg/pull/53735 that currently the rename function is only available to group block, but will be extended to other ‘container’ type block. May I suggest that the gallery block also have this rename function? I think it would help organization better.

Rad to be able to point to this PR where work is already underway to bring this feature to more blocks.

File: packages/block-editor/src/hooks/metadata.js
Comment (getdave) [inline packages/block-editor/src/hooks/metadata.js:56]: This needed to be removed to allow tests to pass otherwise an additional `<div>` was added to output of all blocks.

All functionality and tests appears to work without it. I think it was mistakenly included in the original PR.

**Comment by getdave** [comment]
⚠️ **Do not backport this PR**. 

We will need a Core backport _following_ the release of 6.4 as we do not want this in 6.5.

File: lib/blocks.php
Comment (getdave) [reply lib/blocks.php:476]: I will handle in https://github.com/WordPress/gutenberg/issues/55194

**Comment by youknowriad** [comment]
Hi folks, it looks like this PR introduced a severe performance issue for the "block selection" metric. See https://www.codevitals.run/project/gutenberg

**Comment by getdave** [comment]
Let's roll back now and then take a look at which bit is causing that.

File: packages/block-editor/src/hooks/metadata.js
Comment (getdave) [inline packages/block-editor/src/hooks/metadata.js:26]: I wonder if creating a new object here in an unguarded manner caused the perf regression?

**Comment by getdave** [comment]
I have confirmed offending code is `<BlockSettingsMenuControls>` within `packages/block-editor/src/hooks/block-rename-ui.js`.

**Comment by getdave** [comment]
## Outstanding Work

The following needs to be undertaken in the WP 6.**5** cycle.

- Backport [the refactor (bugfix) to the feature undertaken in the `wp/6.4` branch](https://github.com/WordPress/gutenberg/pull/55604#issuecomment-1779632656)
- [Make "metadata" a global block attribute](https://github.com/WordPress/gutenberg/issues/55194)


------------

**Updated** - I created [an Issue](https://github.com/WordPress/gutenberg/issues/55643)

File: lib/blocks.php
Comment (getdave) [reply lib/blocks.php:476]: @ockham It looks like you backported this to Core in https://core.trac.wordpress.org/changeset/57068.

I was going to say we can remove this code but would that cause backwards compat with versions of WP Core which don't have the backported code but are running the Plugin?

If so it seems to be gated code:

`if ( ! array_key_exists( 'metadata', $args['attributes'] ) ) {`

...so we should probably move to a `compat/wordpress-*` directory? What do you think?

**Comment by getdave** [comment]
@ockham Looks like all the work from this PR was backported to Core already in...

- https://core.trac.wordpress.org/ticket/59797
- https://github.com/WordPress/wordpress-develop/pull/5608

Am I right?

File: lib/blocks.php
Comment (ockham) [reply lib/blocks.php:476]: @getdave I think that makes sense! `compat/wordpress-6.5/` sounds like the right place for this 👍

**Comment by ockham** [comment]
> @ockham Looks like all the work from this PR was backported to Core already in...
> 
> * https://core.trac.wordpress.org/ticket/59797
> * [Block Types: Add `metadata` global attribute wordpress-develop#5608](https://github.com/WordPress/wordpress-develop/pull/5608)
> 
> Am I right?

Yes! AFAICS, the only part that needed backporting was the `metadata` (global) attribute registration, which was covered by the ticket and PR you referenced 👍 

It might however make sense to move the GB "polyfill" from `lib/load.php` to `lib/compat/wordpress-6.5/`, [as you suggested](https://github.com/WordPress/gutenberg/pull/54426/files#r1457353078).

**Comment by youknowriad** [comment]
> It might however make sense to move the GB "polyfill" from lib/load.php to lib/compat/wordpress-6.5/, [as you suggested](https://github.com/WordPress/gutenberg/pull/54426/files#r1457353078).

I think that's a good point and should be done for all the php code that is backported.

**Comment by getdave** [comment]
✅ I updated this PR with the `Backported to Core` label to indicate that [the backport has successfully merged into WP Core](https://github.com/WordPress/gutenberg/pull/54426#issuecomment-1905638331). I also updated the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959).

File: lib/blocks.php
Comment (getdave) [reply lib/blocks.php:476]: I've been looking everywhere for this comment 😅

Comment (getdave) [reply lib/blocks.php:476]: https://github.com/WordPress/gutenberg/pull/58126

**Comment by getdave** [comment]
~PR for move to 6.5 compat dir https://github.com/WordPress/gutenberg/pull/58126~

✅ Completed

**Comment by getdave** [comment]
## Dev note

Building on [the foundation introduced in WordPress 6.4](https://wordpress.org/documentation/wordpress-version/version-6-4/#rename-group-blocks), _all_ blocks (with some exceptions) can now be given custom names within the Editor.

The data for this continues to be stored under the `name` property of the [block's `metadata` attribute](https://github.com/WordPress/wordpress-develop/pull/5608) (e.g. `attributes.metadata.name`).

All blocks (including 3rd party blocks) will be "renameable" **by default**, with the only exceptions being the following blocks:

- core/block
- core/template-part
- core/pattern
- core/navigation
- core/navigation-link

To disable the ability to rename a block, set the relevant supports property to `false` in `block.json`:

```js
// block.json
{
	"supports": {
		"renaming": false // disables ability to rename block via the Editor UI
	},
}
```

**Pull request merged**
